File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        int n = haystack.size();
        int m = needle.size();
        if(n == 1 && m == 1 && haystack[0] == needle[0]){
            return 0;
        }
        for(int i = 0; i<=n-m; i++){
            bool match = true;
            for(int j = 0; j <m; j++){
                if(haystack[i+j] != needle[j]){
                    match = false;
                    break;
                }
            }
            if(match) return i;
        }
        return -1;
        
    }
};

File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if not needle:
            return 0
        n = len(haystack)
        m = len(needle)
        if m > n:
            return -1
        lps = [0]*m
        self.buildlps(needle,lps)
        j = 0
        for i in range (n):
            while j > 0 and haystack[i] != needle[j]:
                j = lps[j-1]
            if haystack[i] == needle[j]:
                j += 1
            if j == m:
               return i - m + 1
        return -1



    def buildlps(self, needle: str, lps: list)-> None:
       m = len(needle)
       length = 0
       i = 1
       lps[0] = 0
       while i < m:
        if needle[i] == needle[length]:
            length += 1
            lps[i] = length
            i += 1
        else: 
            if length != 0:
                length = lps[length-1]
            else:
                lps[i] = 0
                i += 1